src/multiformats/address.cljc

```diff
@@ -2,29 +2,232 @@
+              protocol-key (get code->protocol code)
+              {:keys [codec]} (ensure-protocol-attrs protocol-key)
+              val-start (+ offset code-len)
+              [value num-read] (decode-value codec data val-start)]
+          (cons [protocol-key value]
+                (protocol-value-seq data (+ offset code-len num-read))))))))
+
+
+(defn- concat-arrs [& arrs]
+  (let [total-len (reduce + (map alength arrs))
+        dst (b/byte-array total-len)]
+    (loop [arrs arrs offset 0]
+      (when-let [^bytes src (first arrs)]
+        (b/copy src 0 dst offset (alength src))
+        (recur (rest arrs) (+ offset (alength src)))))
+    dst))
+
+(defn- entry-bytes
+  "build byte array representation for protocol/value pair,
+   note that `value` may be nil for no-value protocols.
+
+   entry can either be a single keyword for a no-value protocol, or a
+   pair of [protocol-key value]. You can also pass in [key nil] for a protocol
+   not expecting a value"
+  [entry]
+  (let [protocol-key (if (keyword? entry) entry (first entry))
+        value (when (coll? entry) (second entry))
+        {:keys [code, codec]} (ensure-protocol-attrs protocol-key)
+        code-bytes (varint/encode code)
+        val-bytes (if (nil? value)
+                    (b/byte-array 0)
+                    (codec/str->bytes codec value))
+        val-len-bytes (if (codec/fixed-byte-length codec)
+                        (b/byte-array 0)
+                        (varint/encode (alength ^bytes val-bytes)))]
+    (concat-arrs code-bytes val-len-bytes val-bytes)))
+
+(deftype Address
+  [^bytes _data _meta ^:unsynchronized-mutable _hash]
+
+  #?(:clj Seqable :cljs ISeqable)
+  (#?(:clj seq :cljs -seq) [this] (protocol-value-seq _data))
+
+  #?(:clj IPersistentCollection)
+
+  #?(:cljs ICounted)
+  (#?(:clj count :cljs -count)
+    [this] (count (seq this)))
+
+  #?(:cljs IEmptyableCollection)
+  (#?(:clj empty :cljs -empty)
+    [this] (Address. (b/byte-array 0) nil 0))
+
+  #?(:cljs IEquiv)
+  (#?(:clj equiv :cljs -equiv) [this other]
+    (and (instance? Address other)
+         (b/bytes= (.-_data this) (.-_data ^Address other))))
+
+  #?(:cljs ICollection)
+  (#?(:clj cons :cljs -conj) [this entry]
+    (Address. (concat-arrs _data (entry-bytes entry)) _meta 0))
+
+  IMeta
+  (#?(:clj meta, :cljs -meta)
+    [this]
+    _meta)
+
+
+  #?(:clj IObj :cljs IWithMeta)
+  (#?(:clj withMeta :cljs -with-meta)
+    [this new-meta]
+    (Address. _data new-meta _hash))
+
+  #?(:cljs IHash)
+  (#?(:clj hashCode, :cljs -hash)
+    [this]
+    (if (zero? _hash)
+      (let [hc (hash (b/byte-seq _data))]
+        (set! _hash hc)
+        hc)
+      _hash))
+
+  Object
+  (toString [this]
+     (->> this
+          (mapcat (fn [[k v]] [(name k) v]))
+          (remove nil?)
+          (str/join "/")
+          (str "/"))))
+
+
+(defn- parse-entries
+  [address-str]
+  (let [parts (->> #"/" (str/split address-str) (remove str/blank?))]
+    (loop [parts parts pairs []]
```

Comment:
You could collapse the `let` above and build the parts list in this loop binding.